include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1136,34 +1146,248 @@ __parallel_set_scan(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range
+            .wait();
+        return __n_diff_1;
+    }
+
+    // Otherwise, merge the sequences together
+    auto __tmp_rng4 = __keep_tmp4(__buf_2, __buf_2 + __n_diff_2);
+    auto __tmp_rng3 = __keep_tmp3(__buf_1, __buf_1 + __n_diff_1);
+
+    oneapi::dpl::__par_backend_hetero::__parallel_merge_impl<__set_symmetric_difference_merge_wrapper<_CustomName>>(
+        __q, __tmp_rng3.all_view(), __tmp_rng4.all_view(), std::forward<_Range3>(__result), __comp)
+        .wait();
+    return __n_diff_1 + __n_diff_2;
+}
+
+template <typename _CustomName, typename _Range1, typename _Range2, typename _Range3, typename _Compare,
+          typename _UseReduceThenScan>
+std::size_t
+__set_write_a_only_op(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __result, _Compare __comp,
+                      oneapi::dpl::unseq_backend::_IntersectionTag, _UseReduceThenScan)
+{
+    if constexpr (_UseReduceThenScan::value)
+        return __parallel_set_reduce_then_scan_set_a_write<_CustomName>(
+                   __q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2), std::forward<_Range3>(__result),
+                   __comp, oneapi::dpl::unseq_backend::_IntersectionTag{})
+            .get();
+    else
+        return __parallel_set_scan<_CustomName>(__q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2),
+                                                std::forward<_Range3>(__result), __comp,
+                                                oneapi::dpl::unseq_backend::_IntersectionTag{})
+            .get();
+}
 
-    if constexpr (_SetTag::__can_write_from_rng2_v)
+template <typename _CustomName, typename _Range1, typename _Range2, typename _Range3, typename _Compare,
+          typename _UseReduceThenScan>
+std::size_t
+__set_write_a_only_op(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __result, _Compare __comp,
+                      oneapi::dpl::unseq_backend::_DifferenceTag, _UseReduceThenScan)
+{
+    if constexpr (_UseReduceThenScan::value)
+        return __parallel_set_reduce_then_scan_set_a_write<_CustomName>(
+                   __q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2), std::forward<_Range3>(__result),
+                   __comp, oneapi::dpl::unseq_backend::_DifferenceTag{})
+            .get();
+    else
+        return __parallel_set_scan<_CustomName>(__q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2),
+                                                std::forward<_Range3>(__result), __comp,
+                                                oneapi::dpl::unseq_backend::_DifferenceTag{})
+            .get();
+}
+
+template <typename _CustomName>
+struct reduce_then_scan_wrapper
+{
+};
+
+template <typename _CustomName>
+struct scan_then_propagate_wrapper
+{
+};
+
+template <typename _SetTag>
+struct __consider_write_a_alg
+{
+    // empirically determined threshold for using write_a algorithm
+    static constexpr std::size_t __threshold = 65536;
```

Comment:
Optional suggestion: Have you tested to see if the size of input / output size types influence this threshold? If it does, it might be useful to have a small heuristic based on that.

Reply:
Thanks for the suggestion. I looked at this and found that due to the temporary register memory requirements of the balanced path, the value type does impact the threshold with larger types favoring set_a write, and smaller favoring balanced path.  

I've also found that the threshold really depends on the size of "set a" in the set operation if we were to use set_a_write algorithm (which is `__rng1` for operations other than union, where set a must be  `__rng2`). 

Therefore, the threshold is now based upon the size of this "set a", and scaled by the value type of that set.